### PR TITLE
AP-567 Add Space Kit Icons

### PR DESCRIPTION
**Pull request by justinanastos**
- [x] Add Space Kit icons (including restrictive license)
- [x] Add TypeScript support
- [x] Add necessary dependencies and scripts to create TypeScript icon components from SVGs
- [x] Add icons to storybook story
- [x] Document how to publish updates to npm

---

I especially need help with the docs. Any and all extra commits to help explain them better would be appreciated.

---

@adamatorres : Does this documentation suffice so you know how to add new icons? Also, is the license sufficient?

~~@trevorblades : I need help figuring out how to publish changes here~~ @hwillson helped me!

Resolves issue [AP-567](https://golinks.io/j/AP-567)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,7 @@
 node_modules
 reset.css
+icons/*.tsx
```
Comment (mayakoneval) [inline .gitignore]: why are the icons not in this repo? Because of the icon license?

Comment (justinanastos) [reply .gitignore]: The SVG icons _are_ in the repo, the generated files are not.

**Comment by justinanastos** [comment]
> looks rad - I will be adding icons here with the new nav design
> 
> so we are dropping [1513](https://github.com/mdg-private/engine-frontend/pull/1513)?

We'll only drop mdg-private/engine-frontend#1513 if we can get this in reasonably quickly. There might be a lot of discussion around the design decisions here.

**Comment by adamatorres** [comment]
Can we take the "waiting" folder out of the directory? They don't need to be in there. And where can I read the license?

**Comment by adamatorres** [comment]
@justinanastos -- Can we take the "waiting" folder out of the directory? They don't need to be in there. And where can I read the license?

**Comment by justinanastos** [comment]
@adamatorres 	

> @justinanastos -- Can we take the "waiting" folder out of the directory? They don't need to be in there. And where can I read the license?

Removed! License is https://github.com/apollographql/space-kit/blob/justin/icons/LICENSE.md#icons and https://github.com/apollographql/space-kit/blob/justin/icons/icons/LICENSE.md

**Comment by justinanastos** [comment]
No idea why netlify stopped building :(

File: .gitignore (hunk 1 above)
Comment (justinanastos) [reply .gitignore]: Sorry, not precise language. The SVGs are under source control, the compiled files are not (they are generated on the fly, we don't care what their history looks like)

**Comment by evans** [comment]
Was looking at the [netlify preview](https://5d0ccbf57326da0008ca7fdd--space-kit.netlify.com/?path=/story/space-kit--icons), curious what the asterisk represents. I tried to hover and didn't see any diagnostics. Some documentation at the top would help me navigate around. A table of contents that jumps me to the icon categories would make parsing the story faster. An ephemeral highlight, like [stack overflow](https://stackoverflow.com/a/48909719), would make it even better to navigate.

image

@justinanastos can you give us some more direction on what code you'd like looked at and the documentation you're interested int(an example would help here)

File: README.md [hunk 2]
```diff
@@ -87,12 +102,12 @@ When you access a color by name (i.e. `colors.indigo`), you'll find a palette of
 **CSS-in-JS**
 
 ```jsx
-import styled from '@emotion/styled';
-import {colors} from '@apollo/space-kit';
+import styled from "@emotion/styled";
+import { colors } from "@apollo/space-kit";
```
Comment (mayakoneval) [inline README.md]: are there docs on how to import / use the icons? I have tried with the beta version and am not able to render anything by doing 

`import { IconBack } from `@apollo/space-kit/icons/IconBack`;` and then `<IconBack />`

File: README.md [hunk 3]
```diff
@@ -48,7 +63,7 @@ You'll probably want to include this file once in your app, ideally at the top-m
 **JS + webpack or similar**
 
 ```js
-import '@apollo/space-kit/reset.css';
+import "@apollo/space-kit/reset.css";
 ```
 
 **LESS**
```
Comment (mayakoneval) [inline README.md]: i know this isnt part of this pr, but the less import string has no closing quote right now

**Comment by daniman** [comment]
@justinanastos I think it would be helpful to give a usage example in the readme. Something as simple as --

Usage:
`import { IconBack } from "@apollo/space-kit/icons/IconBack";`

File: README.md (hunk 2 above)
Comment (justinanastos) [reply README.md]: There's no fill issue, you're not setting a size so the elements are rendering at 0px wide and 0px tall. I apparently left this out of the readme. Maybe we should also have an example and a FAQ like "why isn't anything rendering?" and "how do I import icons"

File: icons/src/scripts/convert.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          .map(async filename => {
+            const svgCode = fs.readFileSync(
+              path.join(fullSvgDirectory, filename),
+              "utf-8"
+            );
+
+            // We have to use a custom svgo setup because the `svgr` one isn't
+            // configurable and will sometimes remove fills and strokes that we
+            // don't want removed.
+            const { data: svg } = await svgo.optimize(svgCode, {
```
Comment (cheapsteak) [inline icons/src/scripts/convert.js]: Could I confirm that this issue is resolved by this custom svgo config?  
https://github.com/mdg-private/engine-frontend/pull/1513#discussion_r296054605

Comment (justinanastos) [reply icons/src/scripts/convert.js]: Great find @cheapsteak . Check out the build preview in Netlify for `IconServices`:

image

**Comment by justinanastos** [comment]
> I tried to hover and didn't see any diagnostics.

@evans Tooltips take too long in web browsers :( Those tooltips show that these are streamline icons and was for my information

File: README.md (hunk 2 above)
Comment (mayakoneval) [reply README.md]: oo thank you! If you want notes on what I did to debug I can keep some.

File: README.md [hunk 5]
```diff
@@ -112,9 +129,69 @@ function MyComponent() {
...
+
+##### Why does each icon have a named export instead of a default export?
+
+The engine-frontend team and Apollo OSS teams have decided to not use default exports; this continues that trend.
+
 ## Developing Space Kit
 
+### Icons
+
+To add new icons, add SVGs to [`./icons/src/svgs`](./icons/src/svgs) and open a pull request. The React components will be generated and the TypeScript will be transpiled automatically after merging to `master`.
```
Comment (mayakoneval) [inline README.md]: instead:

To add new icons, add SVGs to [`./icons/src/svgs/some_folder_name`](./icons/src/svgs) in a folder name that makes sense. The colors of this icon will not be maintained when the React components are generated. All colors that are black will be converted to be configurable and all other colors will be unchangeable when consumed.

**Pull request merged**